Windows Explorer Integration: Added support for other languages & tabs (Win 11) + code refactoring

xavier-ateo opened the pull request:
The file picker GUI for Explorer did not work for me on Windows 11 24H2. I have addressed several issues:

**Multiple language support**
Previously, the Windows Explorer integration was only targeted towards English as the operating system language. It now also supports German, and other languages can be easily added.

**Windows 11 tabs**
Path detection was broken when multiple tabs were open in one Explorer window. This is now fixed for English and German.

**Logging and code refactoring**
Logging can be easily enabled to debug issues and figure out which format a specific language uses when you want to add support for it.

**Todo**
- I took over the previously existing logic for determining OneDrive user folders. However I did not test if it still works as intended.
- I also did not test it under Windows 10.

xavier-ateo committed:
Windows Explorer Integration: Added support for other languages & exporer tabs in Windows 11 + code refactoring

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nriley commented on apps/windows_explorer/windows_explorer.py:
```diff
@@ -1,7 +1,51 @@
 from talon import Context, Module, actions, app, ui
 
+# Configuration
+DEBUG_LOGGING = False
```
In general, we tend to use Talon settings rather than constants in the code to configure logging verbosity.

nriley commented on apps/windows_explorer/windows_explorer.py:
```diff
@@ -1,7 +1,51 @@
+DEBUG_LOGGING = False
+
+# Configure logging
+if DEBUG_LOGGING:
+    logging.basicConfig(level=logging.DEBUG, format="%(levelname)s: %(message)s")
```
If you want to declare a logger that is private to your module you can use something like https://github.com/talonhub/community/blob/14e89d69cfb8b59a0e7f403838c352249fa0c9af/core/command_client/rpc_client/rpc_client.py#L13. Then generally you can just set the log level based on the Talon setting as above and the overhead from constructing log messages will be negligible when compared with the verbosity repeated checks add to your code.

nriley commented on apps/windows_explorer/windows_explorer.py:
```diff
@@ -29,125 +73,326 @@
+        current_language_mapping["folder_names"], get_folder_path, user_display_name
+    )
+
+    directories_to_exclude = [""] + current_language_mapping["excludes"]
```
Can you explain why you have an empty string in this list?

nriley commented on apps/windows_explorer/windows_explorer.py:
```diff
@@ -29,125 +73,326 @@
 class UserActions:
 
     def file_manager_open_parent():
+        """Navigate to parent directory."""
```
In general, we do not put docstrings in a context implementation, just in the module definition of an action.

xavier-ateo replied:
The original implementation also had docstrings. Should I remove all of them?

nriley replied:
Please, as long as there is nothing missing from the module docstring.

nriley committed:
Merge branch 'main' into main-upstream

nriley committed:
Merge branch 'main' into main-upstream

xavier-ateo commented on apps/windows_explorer/windows_explorer.py:
```diff
@@ -29,125 +73,326 @@
+            ],
+            "folder_names": {  # Right side is the localized name
+                "Desktop": "Desktop",
+                "Documents": "Dokumente",
```
You are right, there is a win32 API. I will use that instead of the hardcoded values for the user folder names.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

xavier-ateo committed:
Windows Explorer Integration: Major code restructure with and Windows API integration

Refactoring of previous Windows Explorer integration:

**Improvements:**
- Use Windows API for user folder display names instead of hardcoded values
- Use Windows API for user folder paths. This removes the need for the previous OneDrive user folder detection
- Simplified language mapping structure and removed merging logic
- Reorganized code into logical sections

**Code review adjustments:**
- Use "logger = logging.getLogger(__name__)" for logging
- Removed docstrings from UserActions methods

nriley committed:
Merge branch 'main' into main-upstream

nriley commented on apps/windows_explorer/windows_explorer.py:
```diff
@@ -29,59 +91,380 @@
+    guid = GUID()
+    guid.Data1 = int(guid_string[0:8], 16)
+    guid.Data2 = int(guid_string[8:12], 16)
+    guid.Data3 = int(guid_string[12:16], 16)
```
The Python `uuid.UUID` object can give you these with `.time_low`, `.time_mid` and `.time_hi_version`. I don't see a way to get 8 individual bytes so you'll have to do that yourself!

nriley reviewed (changes requested):
From the community backlog session — apologies for the delay in review, as we did not realize that you had made some additional changes. This is looking really good! Do you want to have us consider merging it soon and make the other changes in a future PR?

nriley committed:
Merge branch 'main' into main-upstream

nriley committed:
Merge branch 'main' into main-upstream
